### ISSUE TITLE
Redux is in violation of WordPress.org plugin guidelines

### ISSUE THREAD

**Issue by Ipstenu** [reduxframework/redux-framework#3650]
We emailed the ONE user with commit access to the plugin on WP.org details about a serious guideline infraction.

If you received this email, please REPLY IMMEDIATELY so we know you're aware and will correct this issue. If you'd did NOT get this email, please email `plugins@wordpress.org` as soon as possible.

**Comment by dovy** [comment]
@Ipstenu, I received that email only 48 hours ago, and my wife went into surgery two days ago, you'll forgive my delay. Also, this is a new requirement for WP.org, so saying this is a `serious guideline infraction` is a bit of a stretch don't you think? It feels like a new policy change. Despite that, we will work to comply as is reasonable.

Please be mindful we recently lost a core member of our team, so development is going slower for this open source project. Know that WP.org only represents a small % of our installed user base. This change will impact millions of sites worldwide. So we would request time to test it adequately.

Lastly, we are heavily focused on our next version of Redux and would prefer to NOT develop anymore on Redux 3. You are welcome to test the new Redux 4, found here. https://github.com/reduxframework/redux-framework-4. One of the items is resolved in that code, though others may appear. We are putting all our efforts into Redux 4, given it has much better coding and security.

We would LOVE you to do a review of the new code. In fact, about 3 months ago I asked if the Plugin team would do a code review to ensure we complied, but I was informed: "the requirements would not change and we do not do code reviews for existing plugins." If memory serves, that was you wasn't it?

Thank you for your time. How would you like to proceed?

**Comment by dovy** [comment]
Also please note, there was nothing nefarious in our decision to offload the JS. It was a request of the userbase to lighten the load and reduce page loads by using a CDN. It's so interesting how we're now reversing that. Would it be acceptable to make a secondary plugin and ask people to install that, or should we just increase our payload size regardless in order to comply? The reason I ask is we created a secondary plugin before but were not allowed to add it to WP.org. Trying to find a graceful middle-ground for people.

**Comment by Ipstenu** [comment]
I'm sorry to hear about your wife! I hope she recovers quickly.

My concern was that on WordPress, you only have one person with commit access, which means there's a single point of failure. Given the usage of the code, we didn't want to have this be another case where the one person who got the email wasn't around. 

We strongly recommend you add someone else as a backup committer to your plugin, if possible.

> Also, this is a new requirement for WP.org, 

This is not a new policy (unless 7 years is 'new'). We've clarified it as of 4 years ago, that offloading data (JS and CSS included) isn't permitted. And we've updated that again last year to explain part of this is related to GDPR.

Basically you're allowing these services to track users **without** their opt-in consent. You can't do that. You've never been able to do that.

>  It was a request of the userbase to lighten the load and reduce page loads by using a CDN.

I hear you :(

The problem is that you're a framework. A normal plugin, we'd say "Make it an OPT IN feature." But for you, the person you need to have opting in isn't the person who includes your framework with their code, but the person who installs that other plugin :(

So ...unless you want to code some wp-admin interface to let people toggle that, and have it set NOT to use the CDN by default, it's got to be removed.

**Comment by dovy** [comment]
Nice side-skirt of my other comments. So given you’re newly enforcing this, are you willing to look at the new code? And yes, we recently lost a core team member.

**Comment by Ipstenu** [comment]
Sorry, I missed the question part of that, to look at the new code!

The only thing that popped up was this:

`/redux-framework-4-master/ReduxCore/inc/classes/class-redux-dev.php:42:						'server_file' => '//reduxframework.com/wp-content/uploads/redux/redux_notice.json',`

Which since I think that ONLY calls if you're in dev mode, is fine.

> Lastly, we are heavily focused on our next version of Redux and would prefer to NOT develop anymore on Redux 3.

That is _perfectly acceptable_ to us. Do you have a general ETA for it? 

I do want to make clear, we're not newly enforcing anything here. We're trying to improve our ability to catch issues as soon as possible, and in doing so found this. As horrific as this is to hear, we don't actively trawl the directory for this sort of thing (mostly for lack of a tool that can scan 80,000 plugins efficiently and regularly).

No one here on .org assumed **anything** nefarious. I get that you feel hurt and rushed and under the gun here, but honestly our push was just to make sure someone actually did get the message, understood why it was a problem, and was going to fix it. We trust you! Things like this happen a lot and they're usually just like this, people trying to solve one big problem and missing that they cased a new one.

You know, human error :) It's okay. No one's mad, and now that we know you're working on it 9which puts you one up on another framework who had a four letter word for us) there's NO deadline. We'd rather it get fixed right and make as many people happy as possible.

**Comment by dovy** [comment]
@Ipstenu I really appreciate the change in your tone. I'm really not trying to be a pain, I want to comply.

- Yes, that notice only appears in dev_mode or when WP_DEBUG is enabled. I am glad that's acceptable. 

As far as an ETA, I'm working on some core backward compatibility issues, but I really can't give a solid ETA. I thought it was ready before I found these issues. I'm really obsessive on not causing issues for legacy people.

We are still calling Ace Editor via CDN though, so I'll have to make that change. So I'll make an issue. I fear this is not my day job, so I work on things as I am able. I hope that makes sense.

**Comment by Ipstenu** [comment]
Ironically this isn't my day job either :) 

Once we know it's being attended to, we try to trust everyone that they'll get it fixed. I'm putting an internal note on the plugin that this will be fixed in v4, and let it ride for now. Once that's out, I'll run a scan on all the plugins (ouch), find everyone NOT on v4, and give them a kick to update their plugins. Which ... well, we try to do that anyway when there are security notices. No big.

**Comment by dovy** [comment]
Sweet. I'll let you all know.

Side note, I asked Otto before but he didn't have any ideas. Is there ANY way to do some basic A/B testing on releases or any strategies you have seen people take? Want to make sure there isn't something bad that I missed.

**Comment by Ipstenu** [comment]
The only one that comes to mind is what Ninja Forms did, but they basically included two versions of the plugin and let the users pick what they wanted. I can't say as that would be very effective for _everyone_ out there :(
